### PR TITLE
Further speedups to can-match

### PR DESCRIPTION
**Pull request by original-brownbear**
Most importantly this one removes spurious synchronization. Also, saves some more allocations and indirections in another sorting spot.

relates to https://github.com/elastic/elasticsearch/issues/109527 but still not the end of possible fixes here

### CODE REVIEW

**Comment by original-brownbear** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java [hunk 1]
```diff
@@ -537,11 +537,24 @@ private static List<SearchShardIterator> sortShards(
         MinAndMax<?>[] minAndMaxes,
         SortOrder order
     ) {
-        return IntStream.range(0, shardsIts.size())
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java:540]: This is an unfortunate implementation to begin with, but it's hard to do this on top of an `int[]` I guess. At least doing it this way is quite a bit faster by not instantiating a tree map under the hood (from `sorted()` and having a lot less function calls.

File: server/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java [hunk 2]
```diff
@@ -448,7 +446,11 @@ private static final class CanMatchSearchPhaseResults extends SearchPhaseResults
         @Override
         void consumeResult(CanMatchShardResponse result, Runnable next) {
             try {
-                consumeResult(result.getShardIndex(), result.canMatch(), result.estimatedMinAndMax());
+                final boolean canMatch = result.canMatch();
+                final MinAndMax<?> minAndMax = result.estimatedMinAndMax();
+                if (canMatch || minAndMax != null) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java:451]: This should help quite a bit in spare scenarios. We probably don't need the minAndMax check here, it's never `null` for `false`, but lets fix that in a follow-up IMO by making this fact a compile time safe thing.

**Review by benwtrent — APPROVED** [review]

**Comment by original-brownbear** [comment]
Thanks Ben!

**Pull request merged**